packages/scm/src/browser/scm-tree-widget.tsx

```diff
@@ -0,0 +1,990 @@
+    groupId: string;
+    path: string;
+    sourceUri: string;
+    children: (FileChangeFolderNode | FileChangeNode)[];
+}
+
+export namespace FileChangeFolderNode {
+    export function is(node: TreeNode): node is FileChangeFolderNode {
+        return 'groupId' in node && 'sourceUri' in node && 'path' in node && 'children' in node;
+    }
+}
+
+export interface FileChangeNode extends SelectableTreeNode {
+    sourceUri: string;
+    decorations?: ScmResourceDecorations;
+}
+
+export namespace FileChangeNode {
+    export function is(node: TreeNode): node is FileChangeNode {
+        return 'sourceUri' in node
+            && !FileChangeFolderNode.is(node);
+    }
+}
+
+export interface ScmTreeProps extends TreeProps {
+    defaultExpansion?: 'collapsed' | 'expanded';
+    nestingThreshold?: number;
+}
+
+@injectable()
+export class ScmTreeWidget extends TreeWidget {
+
+    static ID = 'scm-resource-widget';
+
+    static RESOURCE_GROUP_CONTEXT_MENU = ['RESOURCE_GROUP_CONTEXT_MENU'];
+    static RESOURCE_GROUP_INLINE_MENU = ['RESOURCE_GROUP_INLINE_MENU'];
+
+    static RESOURCE_FOLDER_CONTEXT_MENU = ['RESOURCE_FOLDER_CONTEXT_MENU'];
+    static RESOURCE_FOLDER_INLINE_MENU = ['RESOURCE_FOLDER_INLINE_MENU'];
+
+    static RESOURCE_INLINE_MENU = ['RESOURCE_INLINE_MENU'];
+    static RESOURCE_CONTEXT_MENU = ['RESOURCE_CONTEXT_MENU'];
+
+    @inject(MenuModelRegistry) protected readonly menus: MenuModelRegistry;
+    @inject(CommandRegistry) protected readonly commands: CommandRegistry;
+    @inject(CorePreferences) protected readonly corePreferences: CorePreferences;
+    @inject(ScmContextKeyService) protected readonly contextKeys: ScmContextKeyService;
+    @inject(LabelProvider) protected readonly labelProvider: LabelProvider;
+    @inject(EditorManager) protected readonly editorManager: EditorManager;
+    @inject(DiffNavigatorProvider) protected readonly diffNavigatorProvider: DiffNavigatorProvider;
+    @inject(IconThemeService) protected readonly iconThemeService: IconThemeService;
+
+    protected readonly toDisposeOnRepositoryChange = new DisposableCollection();
+
+    constructor(
+        @inject(TreeProps) readonly props: ScmTreeProps,
+        @inject(TreeModel) readonly model: TreeModel,
+        @inject(ContextMenuRenderer) protected readonly contextMenuRenderer: ContextMenuRenderer,
+        @inject(ScmService) protected readonly scmService: ScmService,
+    ) {
+        super(props, model, contextMenuRenderer);
+        this.id = 'resource_widget';
+    }
+
+    @postConstruct()
+    protected init(): void {
+        super.init();
+        this.addClass('groups-outer-container');
+
+        this.refreshOnRepositoryChange();
+        this.toDispose.push(this.scmService.onDidChangeSelectedRepository(() => {
+            this.refreshOnRepositoryChange();
+            this.forceUpdate();
+        }));
+        this.toDispose.push(this.labelProvider.onDidChange(event => {
```

Comment:
`TreeWidget` already handler `labelProvider`, sub tree widget should use `this.toNodeIcon`, `this.toNodeName` and `this.toNodeDescription`, they should not inject `LabelProvider` and listen to its changes

Reply:
You are correct.  I removed the label provider code from scm-tree-widget and the labels still update.